Source/DafnyCore/PrecedenceLinter.cs

```diff
@@ -0,0 +1,236 @@
+//-----------------------------------------------------------------------------
+//
+// Copyright (C) Microsoft Corporation.  All Rights Reserved.
+// Copyright by the contributors to the Dafny Project
+// SPDX-License-Identifier: MIT
+//
+//-----------------------------------------------------------------------------
+using System.Collections.Generic;
+using System.Linq;
+using System;
+using System.Diagnostics.Contracts;
+using JetBrains.Annotations;
+using Microsoft.Boogie;
+
+namespace Microsoft.Dafny {
+
+  public class PrecedenceLinter : IRewriter {
+    internal override void PostResolve(Program program) {
+      base.PostResolve(program);
+      foreach (var moduleDefinition in program.Modules()) {
+        foreach (var topLevelDecl in moduleDefinition.TopLevelDecls.OfType<TopLevelDeclWithMembers>()) {
+          foreach (var callable in topLevelDecl.Members.OfType<ICallable>()) {
+            var visitor = new PrecedenceLinterVisitor(this.Reporter);
+            visitor.Visit(callable, null);
+          }
+        }
+      }
+    }
+
+    public PrecedenceLinter(ErrorReporter reporter) : base(reporter) {
+    }
+  }
+
+  class LeftMargin {
+    public int Column;
+
+    public LeftMargin(int column) {
+      Column = column;
+    }
+  }
+
+  class PrecedenceLinterVisitor : TopDownVisitor<LeftMargin> {
+
+    private readonly ErrorReporter reporter;
+
+    public PrecedenceLinterVisitor(ErrorReporter reporter) {
+      this.reporter = reporter;
+    }
+
+    protected override bool VisitOneExpr(Expression expr, ref LeftMargin st) {
```

Comment:
It occurs to me that the state variable being a ref is what makes you drive crazy about having to manually having to visit a child. I have the same problem in my PR where I'm just not using the state to propagate indentation because it's a ref.

Would it make it simpler if:
- We keep the state, but you don't use it for leftMargin. State is for gathering stats about the entire tree, not really for top-down traversal anyway.
- We add an in parameter named "context", which you can use for the margin
- We add an out parameter to VisitOneExpr and VisitOneStatement named "childContext", initially set to "context". Setting this makes it possible to select the "context" value for the children

Reply:
The `PrecedenceLinterVisitor` builds on Dafny's general `TopDownVisitor`. During the traversal, the overridable method `VisitOneExpr` performs some action on the given expression `expr` and parameter `st`. The design of `TopDownVisitor` makes it possible for `VisitOneExpr(expr, ...)` to affect the traversal in the following ways:
* It can request that no child of `expr` be visited. This is indicated by returning `false`.
* It can request that the children of `expr` be visited using the same parameter value `st`. This is indicated by returning `true` (and not assigning to `st`).
* It can request that the children of `expr` be visited and provide a new parameter value `st` that will be passed to those children. This is done by returning `true` and setting `st` to a different value.

So, essentially, the traversal passes in one `st` value and gets a `bool` and an `st` back. This is encoded by making the `bool` be the method return value and making `st` a `ref` parameter.

`PrecedenceLinterVisitor.VisitOneExpr` sometimes returns `false`, indicating that the main traversal should not go into the children of `expr`. Those children will therefore not be included in the stats gathered in `st`. Instead, `PrecedenceLinterVisitor.VisitOneExpr` will kick off a new visitation of its children, using a separate `st` object. This is how `PrecedenceLinterVisitor` can let the minimum-left-margin information be computed for just a part of the tree.

The `PrecedenceLinterVisitor` also needs to be able to gather stats for all the nodes that are being visited. The design of `TopDownVisitor` does not use the `ref` parameter to accumulate values across the children; instead, it passes the same `st` to all children of `expr`. For this reason, `PrecedenceLinterVisitor` chooses the type of `st` to be a pointer to an integer.

But `PrecedenceLinterVisitor` never uses the third case above. `PrecedenceLinterVisitor` would therefore be simpler, as you're suggesting, if `st` were not a `ref` parameter. Indeed, if we had another `TopDownVisitor` that streamlined this functionality, we could use it. Such a `TopDownVisitor` would then declare `st` (or `context`, as you're suggesting) as an ordinary in-parameter to `VisitOneExpr`, since the method would only need to return a `bool`.

I suggest we not add another top-down visitor as part of this PR, since a `TopDownVisitor` is not required to return a different `st` value. However, if we ever add another top-down visitor class in the future, then `PrecedenceLinterVisitor` could be changed to build on it instead of on `TopDownVisitor`.

Reply:
Great ! Glad we are on the same wavelength with this. I'm ok if you just convert your comment above in a comment to the class `PrecedenceLinterVisitor`, so that this future refactoring will be easier to identify. Thanks.

Reply:
Good suggestion. Done!